clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -0,0 +1,638 @@
+     * on the preference).
+     * @return Non-empty map of supported protocols and metadata
+     */
+    protected abstract LinkedHashMap<String, ByteBuffer> metadata();
+
+    /**
+     * Invoked when a group member has successfully joined a group.
+     * @param generation The generation that was joined
+     * @param memberId The identifier for the local member in the group
+     * @param protocol The protocol selected by the coordinator
+     * @param memberAssignment The assignment propagated from the group leader
+     */
+    protected abstract void onJoin(int generation,
+                                   String memberId,
+                                   String protocol,
+                                   ByteBuffer memberAssignment);
+
+    /**
+     * Perform synchronization for the group. This is used by the leader to push state to all the members
+     * of the group (e.g. to push partition assignments in the case of the new consumer)
+     * @param leaderId The id of the leader (which is this member)
+     * @param allMemberMetadata Metadata from all members of the group
+     * @return A map from each member to their state assignment
+     */
+    protected abstract Map<String, ByteBuffer> doSync(String leaderId,
+                                                      String protocol,
+                                                      Map<String, ByteBuffer> allMemberMetadata);
+
+    /**
+     * Invoked when the group is left (whether because of shutdown, metadata change, stale generation, etc.)
+     * @param generation The generation that was left
+     * @param memberId The identifier of the local member in the group
+     */
+    protected abstract void onLeave(int generation, String memberId);
+
+
+    /**
+     * Block until the coordinator for this group is known.
+     */
+    public void ensureCoordinatorKnown() {
+        while (coordinatorUnknown()) {
+            RequestFuture<Void> future = sendGroupMetadataRequest();
+            client.poll(future, requestTimeoutMs);
+
+            if (future.failed())
+                client.awaitMetadataUpdate();
+        }
+    }
+
+    /**
+     * Check whether the group should be rejoined (e.g. if metadata changes)
+     * @return true if it should, false otherwise
+     */
+    protected boolean needRejoin() {
+        return rejoinNeeded;
+    }
+
+    /**
+     * Reset the generation/memberId tracked by this member
+     */
+    public void resetGeneration() {
+        this.generation = OffsetCommitRequest.DEFAULT_GENERATION_ID;
+        this.memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
+        rejoinNeeded = true;
+    }
+
+    /**
+     * Ensure that the group is active (i.e. joined and synced)
+     */
+    public void ensureActiveGroup() {
+        if (!needRejoin())
+            return;
+
+        // onLeave only invoked if we have a valid current generation
+        onLeave(generation, memberId);
```

Comment:
@hachikuji I kind of think we should be careful about those terminologies. Otherwise we can easily confuse people. Another example is `doSync` and `performSync`, from the name it is hard to tell the difference. I suggest we do the following for AbstractCoordinator:
1. rename `doSync` to `generateAssignments`, and there is no need to pass in `leaderId` because it is always going to be itself.
2. separate `performSync` to `onElectedAsGroupLeader` and `onEelectedAsGroupFollower`. We can invoke them in JoinGroupResponseHandler.handle()
3. rename `onLeave` to `onGroupJoinStart`
4. rename `onJoin` to `onGroupJoinFinish`
5. rename `sendJoinGroupRequest` to `performGroupJoin`

So the sequence becomes:
1. onGroupJoinStart
2. performGroupJoin
3. onGroupJoinFinish

In performGroupJoin():
1. sendJoinGroupRequest
2. JoinGroupResponseHandler.handle()
    * onAssignedAsGroupLeader (Only valid for leader)
      * generateAssignments
      * sendSyncGroupRequest() with assignments
OR
    * onAssignedAsGroupMember (Only valid for member)
      * sendSyncGroupRequest

3. SyncGroupResponseHandler.handle()
    * getPartitionAssignment and complete the future.

Except for that, the client side code structure looks good to me. I will continue to review the server side code.